### PR TITLE
goss: init at 0.3.16

### LINKED ISSUE #121496: goss
**Project description**
`goss` is a server validation tool. It lets you write assertions in YAML format and runs them really fast to ensure your server matches them, for example it lets you assert some binary is present in path or some command gives an expected result, or that some port is open, etc.
I personally use it a lot to test docker images through its sub-project `dgoss`, for which I'll submit another request.

**Metadata**

* homepage URL: https://github.com/aelsabbahy/goss
* source URL: https://github.com/aelsabbahy/goss
* license: Apache License 2.0 
* platforms: linux. Darwin and windows support is _alpha_ (see the [latest release](https://github.com/aelsabbahy/goss/releases/tag/v0.3.16))

### PR DESCRIPTION
**Pull request by Hyzual**
###### Motivation for this change

Closes #121496

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -4121,6 +4121,12 @@
     githubId = 12491746;
     name = "Masato Yonekawa";
   };
+  hyzual = {
+    email = "hyzual@gmail.com";
+    github = "Hyzual";
+    githubId = 2051507;
+    name = "Joris Masson";
+  };
```
Comment (legendofmiracles) [inline maintainers/maintainer-list.nix:4129]: Adding maintainers is usually in another commit `maintainers: add Joris Masson`

**Review by legendofmiracles — CHANGES_REQUESTED** [review]

**Comment by legendofmiracles** [comment]
Result of `nixpkgs-review pr 121497` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>goss</li>
  </ul>
</details>

File: maintainers/maintainer-list.nix (hunk 1 above)
Comment (Hyzual) [reply maintainers/maintainer-list.nix:4129]: Oops sorry about that 😅. I'll split into two commits. Thanks for the review 😊

File: pkgs/tools/misc/goss/default.nix [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  src = fetchFromGitHub {
+    owner = "aelsabbahy";
+    repo = "goss";
+    rev = "v${version}";
+    sha256 = "1m5w5vwmc9knvaihk61848rlq7qgdyylzpcwi64z84rkw8qdnj6p";
+  };
+
+  vendorSha256 = "1lyqjkwj8hybj5swyrv6357hs8sxmf4wim0c8yhfb9mv7fsxhrv7";
+
+  buildFlagsArray = [ "-ldflags=-s -w -X=main.version=${version}" ];
```
Comment (06kellyjac) [inline pkgs/tools/misc/goss/default.nix:16]:
```suggestion
  preBuild = ''
    buildFlagsArray+=("-ldflags" "-s -w -X main.version=v${version}")
  '';
```

It's arguably better to set `buildFlagsArray` through bash currently, but it's not a big deal, especially for just setting ldflags.
Go packaging docs also don't reflect this suggestion atm.
More details here:
https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/no-flags-array.md

File: pkgs/tools/misc/goss/default.nix [hunk 3]
```diff
@@ -0,0 +1,30 @@
+{ lib, fetchFromGitHub, buildGoModule }:
+
+buildGoModule rec {
+  pname = "goss";
+  version = "0.3.16";
+
+  src = fetchFromGitHub {
+    owner = "aelsabbahy";
+    repo = "goss";
```
Comment (06kellyjac) [inline pkgs/tools/misc/goss/default.nix:9]:
```suggestion
    repo = pname;
```

Can use `pname` here or just leave it as-is

**Review by 06kellyjac — APPROVED** [review]
Looks good, welcome to nixpkgs @Hyzual! :tada: 

I added goss to my nur a couple days ago but I've been lazy in not adding it upstream :sweat_smile: 
https://github.com/06kellyjac/nur-packages/blob/master/pkgs/goss/default.nix

I've made a couple suggestions that you're welcome to implement or disregard :slightly_smiling_face:

File: pkgs/tools/misc/goss/default.nix (hunk 3 above)
Comment (Hyzual) [reply pkgs/tools/misc/goss/default.nix:9]: Nice !

File: pkgs/tools/misc/goss/default.nix (hunk 2 above)
Comment (Hyzual) [reply pkgs/tools/misc/goss/default.nix:16]: Thanks for pointing that out, I'll give `nixpkgs-hammering` a closer look :)

File: pkgs/tools/misc/goss/default.nix [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    description = "Quick and easy server validation";
+    longDescription = ''
+      Goss is a YAML based serverspec alternative tool for validating a server’s configuration.
+      It eases the process of writing tests by allowing the user to generate tests from the current system state.
+      Once the test suite is written they can be executed, waited-on, or served as a health endpoint.
+    '';
+    license = lib.licenses.asl20;
+    platforms = with lib.platforms; linux;
+    maintainers = with lib.maintainers; [ hyzual ];
+  };
```
[the remark this answers is not in the document]
Comment (Hyzual) [reply pkgs/tools/misc/goss/default.nix:29]: I'll add `with lib` to the `meta`. I think I'll leave platforms at linux for now, I have no way of testing if it works for darwin.
Thanks for the offer, I'll add you to the maintainer list :)

**Comment by 06kellyjac** [comment]
Result of `nixpkgs-review pr 121497` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>goss</li>
  </ul>
</details>

**Review by 06kellyjac — APPROVED** [review]
LGTM

Thanks for squashing the changes :+1:

**Comment by Hyzual** [comment]
The `CGO_ENABLED=0` env var was not present, which resulted in a dynamically-linked binary. I added it, so now the build should match the [release script](https://github.com/aelsabbahy/goss/blob/master/release-build.sh) used in goss repository.
I also added the `changelog` meta attribute you had suggested, I had forgotten it :sweat_smile:.

Result of `nixpkgs-review pr 121497` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>goss</li>
  </ul>
</details>

**Comment by 06kellyjac** [comment]
I'm going to be a bit overly verbose with my thoughts here to give you an idea what I was thinking:
Why does it need to be static:
I can understand the goss release has to be static so it can be copied into containers just fine and used on a range of machines but due to the way nixpkgs builds things a dynamically linked binary is essentially static:

```
λ ldd ./result/bin/goss
	linux-vdso.so.1 (0x00007fff8f2da000)
	libpthread.so.0 => /nix/store/v8q6nxyppy1myi3rxni2080bv8s9jxiy-glibc-2.32-40/lib/libpthread.so.0 (0x00007f0c92560000)
	libc.so.6 => /nix/store/v8q6nxyppy1myi3rxni2080bv8s9jxiy-glibc-2.32-40/lib/libc.so.6 (0x00007f0c9239f000)
	/nix/store/v8q6nxyppy1myi3rxni2080bv8s9jxiy-glibc-2.32-40/lib/ld-linux-x86-64.so.2 => /nix/store/v8q6nxyppy1myi3rxni2080bv8s9jxiy-glibc-2.32-40/lib64/ld-linux-x86-64.so.2 (0x00007f0c92583000)
```

I guess we can expect nix users might also want to copy their local goss into containers rather than copying in a static one off github (in which case the `/nix/store` links wont exist)
And I guess the `dgoss` and `kgoss` scripts work off this principal 
https://github.com/aelsabbahy/goss/blob/master/extras/dgoss/dgoss#L77-L78
https://github.com/aelsabbahy/goss/blob/master/extras/kgoss/kgoss#L53-L64

TL;DR in this case it's a good reason to also statically compile, but it's not always necessary to copy the upstream release exactly

Also in this case (and with many go binaries) it's actually slightly smaller when statically compiled because it doesn't really link to much: `10580` before -> `10540` after

**Comment by Hyzual** [comment]
Indeed, I made this change after trying to package `dgoss` which depends on `goss` being statically linked (otherwise we get a `not found` error).

**Comment by Hyzual** [comment]
Thanks for the merge :)
